backend/server/src/db/model/drv.rs

```diff
@@ -0,0 +1,95 @@
+use sqlx::{FromRow, Pool, Sqlite};
+use std::collections::HashMap;
+use tracing::debug;
+
+#[derive(Clone, Debug, FromRow)]
+pub struct Drv {
+    /// Derivation store path
+    pub drv_path: String,
+
+    /// to reattempt the build (depending on the interruption kind).
+    pub system: String,
+}
+
+pub async fn has_drv(pool: &Pool<Sqlite>, drv_path: &str) -> anyhow::Result<bool> {
+    let result = sqlx::query_scalar("SELECT EXISTS(SELECT 1 FROM Drv WHERE drv_path = $1)")
+        .bind(drv_path)
+        .fetch_one(pool)
+        .await?;
+    Ok(result)
+}
+
+/// This will insert a hashmap of <drv, Vec<referrences>> into
+/// the database. The assumption is that the keys are new drvs and the
+/// references may or may not already exist
+pub async fn insert_drv_graph(
+    pool: &Pool<Sqlite>,
+    drv_graph: HashMap<String, Vec<String>>,
+) -> anyhow::Result<()> {
+    let mut reference_map: HashMap<String, Vec<String>> = HashMap::new();
+    // We must first traverse the keys, add them all, then we can create
+    // the reference relationships
+    for (drv_path, references) in drv_graph {
```

Comment:
```suggestion
    for (drv_path, references) in &drv_graph {
```

This will turn `drv_path` and `references` into shared references so you should be able to reuse the same map for both iterations. This is because `HashMap` has [multiple `IntoIter` impls](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#impl-IntoIterator-for-%26HashMap%3CK,+V,+S%3E).